text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,1156 @@
+_RINtNtC3std4iter5ChainINtNtC3std4iter3ZipINtNtC3std3vec8IntoItermEINtNtC3std3vec8IntoItermEEE
+                            5----              5----                    5----
+                          3-----------                                43---------
+                                                                    41--------------------
+                                                                   40-----------------------
+```
+
+The compiler is always supposed to use the longest replacement possible
+in order to achieve the best compression. The compressed symbol looks
+as follows:
+
+```
+_RINtNtC3std4iter5ChainINtB2_3ZipINtNtB4_3vec8IntoItermEBt_EE
+                          ^^^         ^^^               ^^^     back references
+```
+
+Back references have the form `B<base-62-number>_`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The reference-level explanation consists of three parts:
+
+1. A specification of the syntax mangled names conform to.
+2. A specification of the compression scheme.
+3. A mapping of Rust entities to the mangling syntax.
+
+For implementing a demangler, only the first two sections are of
+interest, that is, a demangler only needs to understand syntax and
+compression of names, but it does not have to care about how the
+compiler generates mangled names.
+
+
+## Syntax Of Mangled Names
+
+The syntax of mangled names is given in extended Backus-Naur form:
+
+ - Non-terminals are within angle brackets (as in `<name-prefix>`)
+ - Terminals are within quotes (as in `"_R"`),
+ - Optional parts are in brackets (as in `[<decimal>]`),
+ - Repetition (zero or more times) is signified by curly braces (as in `{<name-prefix>}`)
+ - Comments are marked with `//`.
+
+Mangled names conform to the following grammar:
+
+```
+// The <decimal-number> specifies the encoding version.
+<symbol-name> = "_R" [<decimal-number>] <path> [<instantiating-crate>]
+
+<path> = "C" <identifier>               // crate root
+       | "M" <impl-path> <type>         // <T> (inherent impl)
+       | "X" <impl-path> <type> <path>  // <T as Trait> (trait impl)
+       | "Y" <type> <path>              // <T as Trait> (trait definition)
+       | "N" <ns> <path> <identifier>   // ...::ident (nested path)
+       | "I" <path> {<generic-arg>} "E" // ...<T, U> (generic args)
+       | <backref>
+
+// Path to an impl (without the Self type or the trait).
+// The <path> is the parent, while the <disambiguator> distinguishes
+// between impls in that same parent (e.g. multiple impls in a mod).
+// This exists as a simple way of ensure uniqueness, and demanglers
+// don't need to show it (unless the location of the impl is desired).
+<impl-path> = [<disambiguator>] <path>
+
+// The <decimal-number> is the length of the identifier in bytes.
+// <bytes> is the identifier itself and must not start with a decimal digit.
```

Comment:
It might be a good idea to say something about how the constraint "must not start with a decimal digit" is enforced.

 * While the Rust language itself does not allow identifiers to start with decimal digits, I was  concerned that there might be some way to bypass that and end up with an input that violates this constraint, e.g. via raw-identifers.
 * I eventually realized that raw-identifiers are not allowed to start with a digit (see [RFC 2151](https://github.com/rust-lang/rfcs/blob/master/text/2151-raw-identifiers.md#rationale-and-alternatives)), ...
 * ... and the other potential pitfalls, like `#[export_name]`, do not go through the mangler.

So, it seems that the only known instance of this problem, as [previously noted by @eddyb](https://github.com/rust-lang/rfcs/pull/2603#issuecomment-458410463), is crate names, which can start with a decimal digit.

----

Another reasonable resolution to this issue (and a very general purpose, future-compatible one) is to add an *optional* terminator to the `<decimal-number>` representing the RLE length. @eddyb suggested to me in private discussion the option of saying "`<decimal-number>` may be terminated by a `_`; this terminator is required when `<bytes>` starts with a `_` or a decimal digit."

Reply:
FWIW, for crate names, since they can't be used with leading digit without renaming, it might be a good idea to warn and prepend a `_` so no part of the compiler has to deal with the digit prefix.

Reply:
> (Or if we want the encoding to be unique, we can say "a terminating _ is present if and only if <bytes> starts with a _ or a decimal digit.")

This seems like a good solution and is essentially what @eddyb's implementation already does, IIRC.

Reply:
So do we want to extend this to all identifiers? It would mean every identifier that starts with an underscore ends up using one more byte, but maybe that's not much worse.

Reply:
Seems acceptable to me.